core/src/Cuda/Kokkos_Cuda_ReduceScan.hpp

```diff
@@ -455,126 +454,145 @@ __device__ void cuda_intra_block_reduce_scan(
     const typename FunctorType::pointer_type base_data) {
   using pointer_type = typename FunctorType::pointer_type;
 
-  const unsigned value_count   = functor.length();
-  const unsigned BlockSizeMask = blockDim.y - 1;
-
-  // Must have power of two thread count
-
-  if (BlockSizeMask & blockDim.y) {
-    Kokkos::abort("Cuda::cuda_intra_block_scan requires power-of-two blockDim");
-  }
-
-#define BLOCK_REDUCE_STEP(R, TD, S)              \
-  if (!(R & ((1 << (S + 1)) - 1))) {             \
-    functor.join(TD, (TD - (value_count << S))); \
-  }
-
-#define BLOCK_SCAN_STEP(TD, N, S)                \
-  if (N == (1 << S)) {                           \
-    functor.join(TD, (TD - (value_count << S))); \
-  }
-
-  const unsigned rtid_intra      = threadIdx.y ^ BlockSizeMask;
+  const unsigned value_count = functor.length();
+  const unsigned not_less_power_of_two =
+      (1 << (Impl::int_log2(blockDim.y - 1) + 1));
```

Comment:
How did you come up with that formula?
```C++
for (int i = 0; i < 10; ++i)
  std::cout << i << "  " << (2 << (Kokkos::Impl::int_log2(i - 1) + 1)) << '\n';
```
yields
```\
0  2
1  2
2  4
3  8
4  8
5  16
6  16
7  16
8  16
9  32
```
which I assume is not what you wanted

Reply:
Should I resurrect #4577 ?
Isn't `bit_ceil` what you want?  https://godbolt.org/z/TdPcWeGqY

Reply:
https://godbolt.org/z/1396qsaME works fine for me. Note that I'm doing
```
(1 << (Kokkos::Impl::int_log2(i - 1) + 1))
```
and not
```
(2 << (Kokkos::Impl::int_log2(i - 1) + 1))
```
I'm not quite sure if the single use here justifies introducing another helper function but I wouldn't be surprised to find other places.

Reply:
Duh.  You're right.  Sorry about the noise.

I had been looking into these because they were a bunch of use cases throughout the codebase.  The main issue with the PR was that the intrinsics are not usable in a constexpr context.  If I recall correctly we were debating whether to also have a "fast" version that is not constexpr...